### DISCUSSION TITLE
AssertionError: Insert tlast marker should be done after inserting the FIFOs

### DISCUSSION THREAD

**Discussion by mlsearc** [Xilinx/finn#710] [Q&A]
Hello everyone;
I have a model which is shown in the question below.

As I try to run the command below.
```
test_pynq_board = "ZCU104"
target_clk_ns = 5

from finn.transformation.fpgadataflow.make_zynq_proj import ZynqBuild
model = ModelWrapper(build_dir + "/{}_folded.onnx".format(base_file_name))
model = model.transform(ZynqBuild(platform = test_pynq_board, period_ns = target_clk_ns))
```
I get an error as below;

> ---------------------------------------------------------------------------
> AssertionError                            Traceback (most recent call last)
> <ipython-input-12-8efaa5558577> in <module>
>       4 from finn.transformation.fpgadataflow.make_zynq_proj import ZynqBuild
>       5 model = ModelWrapper(build_dir + "/{}_folded.onnx".format(base_file_name))
> ----> 6 model = model.transform(ZynqBuild(platform = test_pynq_board, period_ns = target_clk_ns))
> 
> /home/best/Desktop/upd_finn/finn/deps/qonnx/src/qonnx/core/modelwrapper.py in transform(self, transformation, make_deepcopy, cleanup)
>     138         model_was_changed = True
>     139         while model_was_changed:
> --> 140             (transformed_model, model_was_changed) = transformation.apply(transformed_model)
>     141         if cleanup:
>     142             transformed_model.cleanup()
> 
> /home/best/Desktop/upd_finn/finn/src/finn/transformation/fpgadataflow/make_zynq_proj.py in apply(self, model)
>     347             dataflow_model_filename = sdp_node.get_nodeattr("model")
>     348             kernel_model = ModelWrapper(dataflow_model_filename)
> --> 349             kernel_model = kernel_model.transform(InsertFIFO())
>     350             kernel_model = kernel_model.transform(GiveUniqueNodeNames(prefix))
>     351             kernel_model.save(dataflow_model_filename)
> 
> /home/best/Desktop/upd_finn/finn/deps/qonnx/src/qonnx/core/modelwrapper.py in transform(self, transformation, make_deepcopy, cleanup)
>     138         model_was_changed = True
>     139         while model_was_changed:
> --> 140             (transformed_model, model_was_changed) = transformation.apply(transformed_model)
>     141         if cleanup:
>     142             transformed_model.cleanup()
> 
> /home/best/Desktop/upd_finn/finn/src/finn/transformation/fpgadataflow/insert_fifo.py in apply(self, model)
>     223                     and final_node.op_type != "IODMA"
>     224                 ):
> --> 225                     assert (
>     226                         final_node.op_type != "TLastMarker"
>     227                     ), """Insert tlast marker should be done
> 
> AssertionError: Insert tlast marker should be done
>                         after inserting the FIFOs
> 

My model is shown as below;

![image](https://user-images.githubusercontent.com/116495049/199571316-d43b2cd3-b178-4228-ac84-b8b319098a20.png)

**Comment by JPPalacios** [comment]
Hi,

Did you ever figure this out? I am also failing that same assertion.

Regards,
